[CodeGen] Compute call frame sizes instead of storing them in MachineBBs

ritter-x2a opened the pull request:
Previously, we would need to update stored call frame sizes whenever a MachineBB with a call sequence is split or call frame instructions are moved. By recomputing them instead, this change avoids the need for keeping track of the call frame sizes throughout transformations. It also fixes several cases where updates of call frame sizes were missing.

The overhead according to the compile time tracker is < +0.03% executed instructions on all CT benchmark categories.

This also allows distinguishing a zero-sized call frame from no open call frame, which is helpful to avoid nested CALLSEQs (which are illegal and cause machine verifier errors).

---

Original Description, under the title "[CodeGen] Distinguish zero-sized call frames from no call frame in MachineBB":
Call frames can be zero-sized. Distinguishing between instructions in a
zero-sized call frame and instructions outside of call frames via the
call frame size info stored in the MachineBBs is helpful to avoid nested
CALLSEQs (which are illegal and cause machine verifier errors).

This patch uses std::optional<unsigned> instead of unsigned to represent
call frame sizes. Locations outside of call frames are marked with
std::nullopt whereas locations in zero-sized call-frames are represented
with a std::optional that contains 0.

ritter-x2a committed:
[CodeGen] Distinguish zero-sized call frames from no call frame in MachineBB

Call frames can be zero-sized. Distinguishing between instructions in a
zero-sized call frame and instructions outside of call frames via the
call frame size info stored in the MachineBBs is helpful to avoid nested
CALLSEQs (which are illegal and cause machine verifier errors).

This patch uses std::optional<unsigned> instead of unsigned to represent
call frame sizes. Locations outside of call frames are marked with
std::nullopt whereas locations in zero-sized call-frames are represented
with a std::optional that contains 0.

ritter-x2a commented:
* **#106965** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/106965?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#106964** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/106964?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ritter-x2a and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

jayfoad commented on llvm/lib/CodeGen/MachineVerifier.cpp:
```diff
@@ -3837,11 +3837,11 @@ void MachineVerifier::verifyStackFrame() {
     }
 
-    if ((int)MBB->getCallFrameSize() != -BBState.EntryValue) {
+    if ((int)MBB->getCallFrameSizeOrZero() != -BBState.EntryValue) {
```
Could you improve the verification here by making EntryValue/ExitValue be std::optionals too, and using the to track the presence/absence of a call frame?

jayfoad replied:
There was no good reason. It was just an oversight. If no tests failed, then I would not have noticed that there was a problem.

ritter-x2a replied:
Alright, I'll look for other such cases then and see if I can fix them. Thanks for the insight!

ritter-x2a replied:
I've updated the PR to a version with the adjusted MachineVerifier and fixes for the lit tests that failed with it. However, it's difficult to say whether this captures all cases where a CALLSEQ might be split across MachineBBs. For instance, I found 126 references (in 62 files) to MachineBasicBlock::splice; I think we'd have to handle them (and other cases, e.g., where new MachineBBs are created) or check that they can't affect CALLSEQs to be really sure we caught everything.
An alternative would be to drop the stored call frame size from the MachineBBs and compute up-to-date call frame sizes only when they are needed, i.e., in the PrologueEpilogueInserter and where I use it in #106965.
What's your opinion on that?

jayfoad replied:
> An alternative would be to drop the stored call frame size from the MachineBBs and compute up-to-date call frame sizes only when they are needed, i.e., in the PrologueEpilogueInserter

The whole point of storing the frame size in `MachineBB` is so that `PEI` does not have to do a complete forward walk over the function to compute frame sizes, in addition to the backward walk that it does to eliminate frame indices. My assumption was that the forward walk would be expensive, but if you can do it without angering https://llvm-compile-time-tracker.com/ then go for it!

ritter-x2a replied:
@jayfoad I just updated this PR to (re)compute the call frame sizes on demand instead of storing them. The compile-time overhead seems to be relatively minor ([see the CT report of the entire PR stack here](https://llvm-compile-time-tracker.com/compare.php?from=34b10e165d809bb133d37dfe934859800f21a100&to=1bd72c4ba6c98ced84b3d7d372dc8a3bc7aee7e9&stat=instructions%3Au)).

ritter-x2a replied:
Ping, @jayfoad

jayfoad reviewed (commented):
> Distinguishing between instructions in a
> zero-sized call frame and instructions outside of call frames via the
> call frame size info stored in the MachineBBs is helpful to avoid nested
> CALLSEQs

Do you have an example of where this is useful?

ritter-x2a commented:
> > Distinguishing between instructions in a
> > zero-sized call frame and instructions outside of call frames via the
> > call frame size info stored in the MachineBBs is helpful to avoid nested
> > CALLSEQs
> 
> Do you have an example of where this is useful?

Yes, PR #106965 (the second one in this stacked PR) relies on this for identifying locations that are inside any call frame (even zero-sized ones) to avoid nested CALLSEQs when thread-local-storage addresses are passed as a function argument.

ritter-x2a committed:
Use std::optional in the MachineVerifier to distinguish zero-sized from no call frames.

ritter-x2a committed:
Catch some cases where the call frame size stored in MachineBasicBlocks was not set or updated.

ritter-x2a committed:
[CodeGen] Compute call frame sizes instead of storing them in MachineBBs

Previously, we would need to update stored call frame sizes whenever a
MachineBB with a call sequence is split or call frame instructions are
moved. By recomputing them instead, this change avoids the need for
keeping track of the call frame sizes throughout transformations. It
also fixes several cases where updates of call frame sizes were missing.

The overhead according to the compile time tracker is < +0.03% executed
instructions on all CT benchmarks.

This also allows distinguishing a zero-sized call frame from no open
call frame, which is helpful to avoid nested CALLSEQs (which are illegal
and cause machine verifier errors).

arsenm commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -282,6 +284,10 @@ class MachineFrameInfo {
   /// It is only valid during and after prolog/epilog code insertion.
   uint64_t MaxCallFrameSize = ~UINT64_C(0);
 
+  /// Call frame sizes for the MachineFunction's MachineBasicBlocks. This is set
+  /// by the MachineFrameSizeInfo constructor and cleared by its destructor.
+  MachineFrameSizeInfo *SizeInfo = nullptr;
```
I don't think this type of ownership makes sense. MachineFrameInfo should be a compilation wide state as the function is processed. This shouldn't be set for one function in one pass that happens to compute it

ritter-x2a replied:
The motivation for that was to make passing this info to `X86TargetLowering::EmitLoweredTLSAddr` less obtrusive in #106965. For this PR here, it's certainly not necessary. I can move this change over to the other PR for now and then we can discuss there how else to pass it to the function (or decide on a different way to avoid the nested CALLSEQs for TLSAddrs).

ritter-x2a committed:
Remove MachineFrameSizeInfo ownership from MachineFrameInfo

Also remove an unintended newline.

ritter-x2a commented:
Ping. I pulled in updates from main and updated the description and title of this PR, since it did have a bit of a shift in subject. Let me know if you'd prefer me to open a new PR for computing instead of storing call frame sizes instead.

ritter-x2a commented:
With PR #106965 now made obsolete by the merged #113706, this PR is no longer necessary to achieve the goal I was after. (Re)-computing instead of storing the call frame sizes (which this PR does) might however still be worthwhile to avoid errors caused by missed updates of call frame sizes.
Let me know if there is interest in still pursuing this PR, otherwise I will close it.

jayfoad commented:
I'm ambivalent. Recomputing frame sizes instead of storing them seems like a cleaner approach, but it also seems like wasted effort in PEI to do a complete walk of the CFG just to recompute them.

ritter-x2a commented:
> I'm ambivalent. Recomputing frame sizes instead of storing them seems like a cleaner approach, but it also seems like wasted effort in PEI to do a complete walk of the CFG just to recompute them.

Then let's shelve this PR for now.
